README.md
# cisco2checkpoint migration tool

Cisco to Checkpoint is a convertion tool developped by GoSecure to help during migration projects. The tool has the ability to convert objects such as networks, services, groups and even firewall rules.

The tool require python 2.7 and a forked version of [ciscoconfparse](https://pypi.python.org/pypi/ciscoconfparse) (included).

Supported input syntaxes are currently IOS and ASA. The script output a file in dbedit syntax.

The script was run on nearly 10 different configurations and lead to many success story but the tool is still experimental. You should use it at your own risk.


## Install

```
git clone --recursive https://github.com/gosecure/cisco2checkpoint c2c
cd c2c
```

## Simple use case: Fresh Install

The easiest and most basic way to convert a cisco configuration file, exported with `show run`, as follow.

```
python2.7 c2c.py --export \
    --ciscoFile 'some_cisco_conf.txt' \
    --syntax asa \
    --policy My_Policy \
    --installOn My_Firewall \
    --output 'network_script.txt' \
```

As a result, the file `network_script.txt` is created and contains the converted policy. Upload the file on the SmartCenter server. Note that you will need a user with a bash shell to upload using SSH. To do so, run `chsh -s /bin/bash` and then logout to apply the change.

Before import, make sure that no "write mode" session is opened with SmartDashboard. Also make sure that there is **no empty lines** and that the return characters are **\n** (not \r\n). Don't forget to **take a backup or a DRC (Database Revision Control)**.

To perform the import, run:

```
dbedit -local -f network_script.txt -ignore_script_failure -continue_updating
```

## Advanced used case: Adding FW to existing policy

If a Checkpoint policy already exist on the SmartCenter server, the script must be aware of the existing objects. It is possible to export those objects in xml format and specify them using a new argument.

First run the following on the SmartCenter.

```
echo "printxml network_objects" > printxml_netobj.txt
echo "printxml services" > printxml_services.txt
echo '<a>' > customer_network_objects.xml
dbedit -local -f printxml_netobj.txt >> customer_network_objects.xml
echo '</a>' >> customer_network_objects.xml
echo '<a>' > customer_service_objects.xml
dbedit -local -f printxml_services.txt >> customer_service_objects.xml
echo '</a>' >> customer_service_objects.xml
```

Then copy both xml files in the root of the repository and run as follow.

```
python2.7 c2c.py --export \
    --ciscoFile 'some_cisco_conf.txt' \
    --cpPortsFile 'customer_service_objects.xml' \
    --cpNetObjFile 'customer_network_objects.xml' \
    --syntax asa \
    --policy My_Policy \
    --installOn My_Firewall \
    --output 'network_script.txt' \
```

As a result, the file `network_script.txt` is created and contains the converted policy. Upload the file on the SmartCenter server. Note that you will need a user with a bash shell to upload using SSH. To do so, run `chsh -s /bin/bash` and then logout to apply the change.

Before import, make sure that no "write mode" session is opened with SmartDashboard. Also make sure that there is **no empty lines** and that the return characters are **\n** (not \r\n). Don't forget to **take a backup or a DRC (Database Revision Control)**.

To perform the import, run:

```
dbedit -local -f network_script.txt -ignore_script_failure -continue_updating
```

## Verify the convertion

Simply replace the `--export` argument by `--verify` and add `--format text` to generate the configuration into a human-readable format. For example:

```
python2.7 c2c.py --verify \
    --format text \
    --ciscoFile 'some_cisco_conf.txt' \
    --syntax asa \
    --policy My_Policy \
    --installOn My_Firewall \
    --output 'network_script_verify.txt' \
```

in the example below, the last three lines shows three `access-list` that were used to generate a single checkpoint rule.

```
ACLRule(name=acl_inside,src=N-Prod-Wks-10.12.160.0_19;N-LAN-Wks-10.16.160.0_24;N-Prod-Wks-10.21.60.0_22,dst=G-xmpp-Internet,port=any,action=permit,pol=FW-Temp,inst=,disabled=False,desc=Access to X)
 Desc:Access to X
 Src: CiscoNet(name=N-Prod-Wks-10.12.160.0_19,ipAddr=10.12.160.0/255.255.224.0,desc=,alias=)
 Src: CiscoNet(name=N-LAN-Wks-10.16.160.0_24,ipAddr=10.16.160.0/255.255.255.0,desc=,alias=)
 Src: CiscoNet(name=N-Prod-Wks-10.21.60.0_22,ipAddr=10.21.60.0/255.255.252.0,desc=,alias=N-Prod-Wkstn-10.21.60.0_22)
 Dst: CiscoNetGroup(name=G-xmpp-Internet,desc= description Routes de xmpp via Internet,nbMembers=6,alias=)
   CiscoNet(name=N-xmpp-Int-160.43.250.0_24,ipAddr=160.43.250.0/255.255.255.0,desc=,alias=)
   CiscoNet(name=N-xmpp-Int-206.156.53.0_24,ipAddr=206.156.53.0/255.255.255.0,desc=,alias=)
 Port: CiscoAnyPort(name=any,port=0,desc=,alias=)
 Verify: <ASAAclLine # 2531 'access-list acl_inside extended permit ip object N-Prod-Wks-10.12.160.0_19 object-group G-xmpp-Internet log'>
 Verify: <ASAAclLine # 2532 'access-list acl_inside extended permit ip object N-LAN-Wks-10.16.160.0_24 object-group G-xmpp-Internet log'>
 Verify: <ASAAclLine # 2534 'access-list acl_inside extended permit ip object N-Prod-Wkstn-10.21.60.0_22 object-group G-xmpp-Internet log'>
```


## Implicit Behaviors

**Firewall rationalization**

By default, firewall rules are merged together if it doesn't affect the security of the policy. For two rules to be merged, all fields except one must be identical. Less firewall rules result in better performance and ease of management, which often increase security.

**Objects reuse**

When two objects with the same properties are found, they are merged. For example, if two hosts are defined with the same IP address, the first name is taken.

**Dynamic Object generation**

When a host is defined in a group or an access-list without a name, it is automatically created. The reason is that Checkpoint everything to be defined object.


## Modifiers

Those modifiers can be used for specific needs, such as specifying color on objects and enable logging on every rules.


    --color COLOR         The color to use for new objects.
    --force-log           Force track=Log on all firewall rules
    --startIndex INDEX    Index to start importing firewall rules. Default: 0
    --disableRules        Disable all firewall rules.
    --flattenInlineNetGroups
                          Flatten groups with prefix DM_INLINE_NETWORK_ so
                          members are added to firewall rules instead of the
                          group.
    --flattenInlineSvcGroups
                          Flatten groups with prefix DM_INLINE_SERVICE_ so
                          members are added to firewall rules instead of the
                          group.


## Customization

A config file is located in `./config.py`. 95% of the users shouldn't need to modify it. However, it is worth knowing the existence of this file. User will find object prefix for dynamically created objects, excluded checkpoint services, illegal expressions replacements and association tables for ports and protocols.


## Warning

**Layer 2 vs Layer 3 :** As some of you may know, Checkpoint policy is layer 3 based, meaning that firewall interfaces are completely abstracted from the policy. In fact, it is not possible to assign a rule to an interface like Cisco do. For this reason, you will find many *Drop All* rules within the policy and other rules that makes no sense in the context of a layer-3 stateful firewall. Thus, a review must **always** be performed after the import.

**Cisco syntax :** Cisco tends to be very flexible when it's time to write a line. Some keywords are optional and it is not required to define everything as an object. It also  allows one to define an object in multiple ways. Unfortunatly, this lead to little hacks in the code.


## License

cisco2checkpoint is licensed GPLv3; Copyright [GoSecure](https://gosecure.net), 20015-2016.


## Author and Thanks

cisco2checkpoint was developped by Martin Dubé (mdube at gosecure.ca)

Special thanks:
 - David Michael Pennington for his awesome work on [ciscoconfparse](https://github.com/mpenning/ciscoconfparse)
 - Olivier Bilodeau for helping me rebase the project and follow good practices.


releases.txt
TODO 
 - Add a "--no-any-any" feature to avoid import of any/any rules
 - Look for examples with client DATA
 - add counters to c2c.addObj() to automatically increase counters
 - check group merge with -01
 - run pep8
 - run pylint
 - run sphinx

v1.1 - 2016-09-07
Features:
 - Clean the code for publishing
 - Extracted changes made in ciscoconfparse to fit in a single file.
 - ciscoconfparse is not a submodule to https://github.com/gosecure/ciscoconfparse
 - License is now GPLv3 to match ciscoconfparse project.

v1.0 - 2016-02-21
Features:
 - Major refactoring of the parsing to reduce "hacks" and ugly codes
 - A distinction is made between ASA and IOS. The argument --syntax was added.
 - Unfortunatly, ciscoconfparse was forked but a pull request to the Project
   should be made in a near future.
 - Added support for source ports
 - Deleted an important number of line of codes

v0.18 - 2016-01-13
Features:
 - Added support to protocols: ospf,esp,ah,ahp,vrrp,skip,gre
 - Added support to "ip access-list"
 - Reduced ACL parsing code from cisco2checkpoint.py
 - Optimized parsing by updating ciscoconfparse library
 - Now using ciscoconfparse with factory=True

v0.17 - 2015-10-13
Features:
 - Added support to import only new objects or flag differences.
 - New parameter: --new-only 
 - New parameter: --show-diff
 - Updated network_objects.xml and services_objects.xml using "printxml services" and "printxml network_objects" in dbedit.
 
v0.16 - 2015-06-22
Features:
 - Developped a feature to "flatten" "DM_INLINE" groups in a firewall rule. When enabled, all these groups are removed and only the member is added to a firewall rule attribute (either src, dst or port). 
 
v0.15 - 2015-06-16
Features:
 - Several adjustement for another customer.
 - Uncommented "names" 
 - Added rtsp port (554)
 - Added some ICMP messages
 - Added tftp and other cisco protocols
 - Added the port object (can parse SinglePort and PortRange) from cisco config
 - Added prefix to CiscoRange when created by script
 - Fixed NAT issues when importing hide nat rules
 
v0.14 - 2015-05-08
Features:
 - Made several tests
 - Fixed a bug where NAT rules were not imported properly in "directory mode" (--configDir)
 - Fixed a bug where groups were not renamed when starting with a number.
 - Fixed a bug where disabled rules were merged with enabled one
 
v0.13 - 2015-05-07
Features:
 - Fixed a bug where groups were not renamed properly 
 - Fixed a bug where "any" objects were not appended to group members
 - Fixed a bug where fw rules did not contain the right objects
 - Fixed a bug where "tcp-udp" protocol was not created properly
 
v0.12 - 2015-04-27
Features:
 - updated formating
 - Removed labels
 - Fixed a bug where group renaming was not working properly

v0.11 - 2015-04-24
Features:
 - Do not merge rules from different ACLs
 - Fixed issues with group renaming
 
v0.10 - 2015-04-21
Features:
 - Fixed NAT issue where "interface" become a hide rule
 - Written Cisco2CheckpointManager class that can interact with multiple config files
 
v0.09 - 2015-04-20 
Features:
 - Worked on a procedure
 - Fixed the "network" alias on CiscoHost
 - Updated text output

v0.08 - 2015-04-17
Features:
 - Fixed several redundancy bugs with "any" objects
 - Fixed a bug where "dest-unreach" and "echo" icmp messages were not imported properly
 - Moved some dictionaries to config.py
 - The script now support policy and installOn arguments
 - Fixed a bug where a mask was imported as subnet

v0.07 - 2015-04-16
Features:
 - Fixed multiple duplicate ports
 - Fixed multiple duplicate source
 - Fix a bug where "any4" is created as an object
 - Fixed a bug where "host" and "network" identifier were imported as objects

v0.06 - 2015-04-15
Features:
 - Fixed bug with comments
 - Fixed a bug where a CiscoRange was imported as a CiscoName
 - Fixed a bug where a CiscoHost was sometimes added as a CiscoNet
 - Removed some prefix/suffix characters in the comment fields
 - Added _fixDuplicateSubnet() to delete duplicate subnets
 - Added _fixDuplicateRange() to delete duplicate subnets
 - Changed order of _fixDuplicateName() because it was causing issues.
 - Added support for nat rules (static only)
 - Fixed a bug where "any" port was not flagged as "alreadyExist"
 - Fixed a bug where some character crash the import if a comments field contain bad chars (such as :)
 - Added a function to merge redundant rules
 
v0.05 - 2015-04-14
Features:
 - Fixed several issues
 - Added a function to sanitize illegal words and characters
 - Added the "any" host object and service object
 - Added a way to print all objects
 
v0.04 - 2015-04-13
Features:
 - Migrate a FW rule
 - Fixed several issues
 - Added a function to import all checkpoint ports in XML format.

v0.03 - 2015-04-09
Features:
 - Migrate a port, port-group
 
v0.02 - 2015-04-09
Features:
 - Migrate a host, subnet, range, network-group
 
v0.01 - 2015-04-08
Features:
 - Migrate a "name"
 


c2c.py
#!/usr/bin/python2
# -*- coding: utf-8 -*-

'''
@author: Martin Dubé
@organization: GoSecure
@contact: mdube@gosecure.ca
@license: GPLv3
Copyright (c) 2016, GoSecure

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation version 3.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

# Python version validation
import sys
if sys.version_info > (2,7,13):
    print('Python version 2.7 is needed for this script')
    exit(1);

sys.path.insert(0, 'lib')
del sys

# Project imports
from cisco2checkpoint import Cisco2Checkpoint,Cisco2CheckpointManager
from config import *

# System imports
import argparse
import os

# Get args
description = '''GoSecure cisco2checkpoint migration tool.

Known limitations
 - The script convert "permit" to "allow" and "deny" to "deny".
 - The script support only None and Log tracking.
 - The script does not feed the "VPN" field.
 - The NAT rule translation is buggy (for now).
 - The script does not support IPv6 (so any4 or any6 become any)
 - Redundant groups are not merged yet.
'''
epilog = '''
Examples:
Print a summary of what is parsed
  ./cisco2checkpoint.py --ciscoFile 'cisco-run-conf.txt' --summary

Search some objects
  ./cisco2checkpoint.py --ciscoFile 'cisco-run-conf.txt'
          --search 'obj-172.16.66.0' --format text
  ./cisco2checkpoint.py --ciscoFile 'cisco-run-conf.txt'
          --search 'obj-172.16.66.0' --format text --filter CiscoHost

Export in a human readable form
  ./cisco2checkpoint.py --ciscoFile 'cisco-run-conf.txt' --export --format text
  
Export for dbedit
  ./cisco2checkpoint.py --ciscoFile 'cisco-run-conf.txt' --export
          --format dbedit --policy Standard --installOn fw01 --installOn fw02

'''
parser = argparse.ArgumentParser(
                    formatter_class=argparse.RawDescriptionHelpFormatter, \
                    description=description, epilog=epilog)
parser.add_argument('-v','--version', action='version', \
                    version='%(prog)s '+C2C_VERSION)
parser.add_argument('--debug', action='store_true', dest='debug', \
                    default=False, help='Run the tool in debug mode')

actGrp = parser.add_argument_group("Action", "Select one of these action")
actGrp.add_argument('--summary', '-u', action='store_true', dest='summary', 
            default=False, \
            help='Print a summary of what is parsed and what would be migrated.')
actGrp.add_argument('--export', '-e', action='store_true', dest='export', default=False, \
              help='Export configuration. Use --format to determine format.')
actGrp.add_argument('--verify', action='store_true', dest='verify', default=False, \
              help='Export configuration like --export but in text format and in a verifyable format.')
actGrp.add_argument('--search', '-s',  action='store', dest='search', default='',\
              type=str, metavar='TEXT', help='Search for a specific object.')

cnfGrp = parser.add_argument_group("Import config")
cnfGrp.add_argument('--ciscoFile', '-c', action='store', dest='ciscoFile', default='', \
              type=str, metavar='FILE', help='Cisco config file to parse.')
cnfGrp.add_argument('--ciscoDir', '-d', action='store', dest='ciscoDir', default='', \
              type=str, metavar='DIR', help='config directory to parse. Will read only *.Config files')
cnfGrp.add_argument('--cpPortsFile', action='store', dest='cpPortsFile', default=DEFAULT_CP_PORT_FILE, \
              type=str, metavar='FILE', help='Checkpoint xml port file to parse. Default: %s' % DEFAULT_CP_PORT_FILE)
cnfGrp.add_argument('--cpNetObjFile', action='store', dest='cpNetObjFile', default=DEFAULT_CP_NETOBJ_FILE, \
              type=str, metavar='FILE', help='Checkpoint xml network objects file to parse. Default: %s' % DEFAULT_CP_NETOBJ_FILE)              
cnfGrp.add_argument('--syntax', action='store', dest='syntax', default=DEFAULT_SYNTAX, \
              type=str, help='Specify the cisco syntax. Valid values: ios, asa. '\
                    'Default: %s' % DEFAULT_SYNTAX)

optGrp = parser.add_argument_group("Options")
optGrp.add_argument('--format', '-f', action='store', dest='format', default=DEFAULT_FORMAT, \
              type=str, help='Specify the format. Valid values: dbedit, text. Default: %s' % DEFAULT_FORMAT)
optGrp.add_argument('--output', '-o', action='store', dest='output', default=DEFAULT_OUTPUT_FILE, \
              type=str, metavar='FILE', help='Output file. Default: %s' % DEFAULT_OUTPUT_FILE)
optGrp.add_argument('--filter', action='append', dest='filter', default=None, \
              type=str, metavar='CLASS', help='Filter a class name, e.g. CiscoHost, CiscoPort, CiscoFwRule. Can use option several times.')
optGrp.add_argument('--stdout', action='store_true', dest='stdout', default=False, \
              help='Print output to stdout.')

expGrp = parser.add_argument_group("Export Modifiers")
expGrp.add_argument('--policy', action='store', dest='policy', default=DEFAULT_POLICY, \
              type=str, help='The policy name. Relevant with --export only. Default: %s' % DEFAULT_POLICY)
expGrp.add_argument('--installOn', action='store', dest='installOn', default=DEFAULT_INSTALLON, \
              type=str, metavar='FWs', help='Specify the checkpoint object to install rules on. ')
expGrp.add_argument('--natInstallOn', action='store', dest='natInstallOn', default=DEFAULT_NAT_INSTALLON, \
              type=str, metavar='FW', help='The firewall to use for all hide and static NAT rules.')
expGrp.add_argument('--color', action='store', dest='color', default=DEFAULT_NEW_OBJ_COLOR, \
              type=str, metavar='COLOR', help='The color to use for new objects.')
expGrp.add_argument('--force-log', action='store_true', dest='forceLog', default=False, \
              help='Force track=Log on all firewall rules')
expGrp.add_argument('--startIndex', action='store', dest='startIndex', default=DEFAULT_ACL_RULE_INDEX, \
              type=int, metavar='INDEX', help='Index to start importing firewall rules. Default: %i' % DEFAULT_ACL_RULE_INDEX)
expGrp.add_argument('--disableRules', action='store_true', dest='disableRules', default=False, \
              help='Disable all firewall rules.')
expGrp.add_argument('--flattenInlineNetGroups', action='store_true', dest='flattenInlineNetGroups', default=False, \
              help='Flatten groups with prefix DM_INLINE_NETWORK_ so members are added to firewall rules instead of the group.')
expGrp.add_argument('--flattenInlineSvcGroups', action='store_true', dest='flattenInlineSvcGroups', default=False, \
              help='Flatten groups with prefix DM_INLINE_SERVICE_ so members are added to firewall rules instead of the group.')

args = parser.parse_args()

if args.debug:
    print(args)

# Step 1: Validate user input
if not os.path.isfile(str(args.cpPortsFile)):
    print('Cannot find checkpoint port file: "%s"' % args.cpPortsFile)
    exit(1)
elif not ((args.ciscoFile and os.path.isfile(str(args.ciscoFile))) \
  or (args.ciscoDir and os.path.isdir(args.ciscoDir))): # if dir was specified
    print('You must specify either a file or a dir (see --ciscoFile and --ciscoDir)')    
    exit(1)
if not (args.syntax in ['ios','asa']):
    print('Invalid --syntax value. RTFM.')
    exit(1)

if args.ciscoFile != '':
    c2c = Cisco2Checkpoint()
elif args.ciscoDir != '':
    c2c = Cisco2CheckpointManager()

c2c.setDebug(args.debug)
c2c.setSyntax(args.syntax)
c2c.setPolicy(args.policy)
c2c.setInstallOn(args.installOn)
c2c.setNatInstallOn(args.natInstallOn)
c2c.setColor(args.color)
c2c.setForceLog(args.forceLog)
c2c.setDisableRules(args.disableRules)
c2c.setACLRuleIndex(args.startIndex)
c2c.setFlattenInlineNetGroups(args.flattenInlineNetGroups)
c2c.setFlattenInlineSvcGroups(args.flattenInlineSvcGroups)

if args.ciscoFile != '':
    c2c.importConfig(args.cpPortsFile,args.cpNetObjFile,args.ciscoFile)
elif args.ciscoDir != '':
    c2c.importConfig(args.cpPortsFile,args.cpNetObjFile,args.ciscoDir)

# Step 2: Process user request
#try:
if args.summary:
    print(MSG_PREFIX+'Generate Summary')
    print(c2c.getSummary())
elif args.search != '':
    if args.filter == None:
        print(MSG_PREFIX+"Searching for an object. No filter")
        obj_list = c2c.findObjByName(args.search)
        if len(obj_list) > 0:
            print(''.join([obj.toString() for obj in obj_list]))
        else:
            print(MSG_PREFIX+'No object found')
    else:
        print(MSG_PREFIX+"Searching for an object with filter(s): %s" % ','.join(args.filter))
        obj_list = c2c.findObjByNameType(args.search,args.filter)
        if len(obj_list) > 0:
            print(''.join([obj.toString() for obj in obj_list]))
        else:
            print(MSG_PREFIX+'No object found')
elif args.export:
    result = ''
    if args.filter == None:
        if args.format == 'text':
            print(MSG_PREFIX+'Exporting to text format')
            result = c2c.getAllObjs()
        elif args.format == 'dbedit':
            print(MSG_PREFIX+'Exporting to dbedit format')
            result = c2c.toDBEdit()
        else:
            print(WARN_PREFIX+'Invalid format')
    else:
        if args.format == 'text':
            print(MSG_PREFIX+'Exporting to text format')
            obj_list = c2c.findObjByType(args.filter)
            if len(obj_list) > 0:
                result += ''.join([obj.toString() for obj in obj_list])
            else:
                print(MSG_PREFIX+'No object found')
        elif args.format == 'dbedit':
            print(MSG_PREFIX+'Operation not supported yet')
        else:
            print(WARN_PREFIX+'Invalid format')
    
    # Print summary
    print(c2c.getSummary())

    # Output
    if args.stdout:
        print(result)
    else:
        fd = os.open(args.output, os.O_RDWR|os.O_CREAT|os.O_TRUNC)
        os.write(fd,result)
        os.close(fd)
elif args.verify:
    result = ''
    args.format = 'text'
    
    if args.filter == None:
        print(MSG_PREFIX+'Exporting to verify format')
        result = c2c.getAllObjs(True)
    else:
        print(MSG_PREFIX+'Exporting to verify format')
        obj_list = c2c.findObjByType(args.filter)
        if len(obj_list) > 0:
            result += ''.join([obj.toString('', True) for obj in obj_list])
        else:
            print(MSG_PREFIX+'No object found')

    # Print summary
    print(c2c.getSummary())

    # Output
    if args.stdout:
        print(result)
    else:
        fd = os.open(args.output, os.O_RDWR|os.O_CREAT|os.O_TRUNC)
        os.write(fd,result)
        os.close(fd)
else:
    parser.print_help()
